### PR TITLE
FIRM16-5: Timer Driver

### PR DESCRIPTION
**Pull request by cam-rod**
PR to merge TIM driver into main, with all suggestions implemented.

### CODE REVIEW

**Comment by cam-rod** [comment]
Updated the TIM driver files to use better directives, and updated pins in tim_test.c to also work on STM32H7. tim_test.c also uses the GPIO driver now.

File: Manual_Tests/STM32_tim/tim_test.c [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * In this context, a time unit is 150 ms (20/6 Hz), so the entire message (54 time units) is transmitted in 8.1 seconds.
+ * Update events occur at each other time unit (300 ms or 20/3 Hz).
+ */
+int main(){
+	setup_tim();
+#ifdef STM32G474xx
+	gpio_init_output(GPIOA, 5);
+#elif defined(STM32H743xx)
+	gpio_init_output(GPIOB, 7);
+#endif
```
Comment (brunofalmeida) [inline Manual_Tests/STM32_tim/tim_test.c:62]: 1. I believe the second arguments should be `GPIO_PIN_5`, equal to `1 << 5`, and `GPIO_PIN_7`, equal to `1 << 7`.

2. I would recommend you refactor this so that the `#ifdef`s for the G4 and H7 are near the top of the file and define some common constants since those are used in multiple places in the file.

Could be something similar to:
```
#ifdef STM32G474xx
    #define LED_PORT GPIOA
    #define LED_PIN GPIO_PIN_5
 #elif defined(STM32H743xx)
    #define LED_PORT GPIOB
    #define LED_PIN GPIO_PIN_7
  #endif
```

Then you won't need `#ifdef`s anywhere else in the file.

3. I would also recommend you save the struct returned from this function into some sort of global variable, so then you can toggle it from your IRQ function.

File: Manual_Tests/STM32_tim/tim_test.c [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+						  1,0,0,1,1,1,0,0,
+						  1,0,0,1,1,1,1,0,0,1,1,1,0,0,
+						  1,1,1,1,1,1,1,1,0,0,0,0,0,0}; // Each line is a letter and the trailing space or end of word.
+
+// Interrupt handler, advances at the end of each period
+void TIM2_IRQHandler(){
+	if(tim_get_IT_flag(&(timer.handle), TIM_IT_UPDATE)){
+		tim_clear_IT_flag(&(timer.handle), TIM_IT_UPDATE);
+		if(led_toggle[time_unit]){
+			HAL_GPIO_TogglePin(GPIOA, GPIO_PIN_5);
```
Comment (brunofalmeida) [inline Manual_Tests/STM32_tim/tim_test.c:32]: Preferably call the GPIO library to do this instead of the HAL directly. If there is no toggle function in the GPIO library, you should ask Brytni about adding one.

File: Src/drivers/STM32_tim/tim.c [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+	}
+}
+
+/* Stops a given timer
+ *
+ * @param TimFunc *timer - the timer struct
+ * @return HAL_StatusTypeDef - if the operation inputs were valid (good stop is HAL_OK or 0x00U)
+ */
+HAL_StatusTypeDef tim_stop(TimFunc *timer){
+	HAL_StatusTypeDef successful = 1;
```
Comment (brunofalmeida) [inline Src/drivers/STM32_tim/tim.c:333]: What does 1 mean in this case? Is it an error? Should use a constant here.

File: Src/drivers/STM32_tim/tim.c [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+		period = (clk_frequency / target_frequency / (prescaler+1)) - 1;
+		qs_timer = tim_config_base(TIM4, mode, prescaler, period);
+	} else {
+		prescaler = (clk_frequency / target_frequency) - 1;
+		period = (clk_frequency / target_frequency / (prescaler+1)) - 1;
+		qs_timer = tim_config_base(TIM4, mode, prescaler, period);
+	}
+
+	HAL_Init();
+	clock_init();
```
Comment (brunofalmeida) [inline Src/drivers/STM32_tim/tim.c:460]: Why are these here? Your timer library shouldn't be handling any of the HAL and clocks initialization.

File: Src/drivers/STM32_tim/tim.h [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+// Constants for timer functions and modes
+enum tim_mode_constants {TIM_Mode_None, TIM_Mode_IT, TIM_Mode_DMA};
+enum tim_func_constants {TIM_Base, TIM_IC, TIM_OC, TIM_PWM, TIM_OnePulse, TIM_Encoder};
+
+// Helper calculator macros
+#ifdef STM32G474xx
+#define tim_calc_prescaler(tim_clk, count_clk) 				__HAL_TIM_CALC_PSC(tim_clk, count_clk)
+#define tim_calc_period(tim_clk, prescaler, frequency) 		__HAL_TIM_CALC_PERIOD(tim_clk, prescaler, frequency)
+#endif
```
Comment (brunofalmeida) [inline Src/drivers/STM32_tim/tim.h:65]: These are only available for the G4. Will it compile and work correctly for the H7?

File: Startup/STM32H7/stm32h7xx_hal_conf.h [hunk 6]
```diff
@@ -213,7 +213,7 @@
 #define  USE_HAL_SMBUS_REGISTER_CALLBACKS   0U /* SMBUS register callback disabled   */
 #define  USE_HAL_SPI_REGISTER_CALLBACKS     0U /* SPI register callback disabled     */
 #define  USE_HAL_SWPMI_REGISTER_CALLBACKS   0U /* SWPMI register callback disabled   */
-#define  USE_HAL_TIM_REGISTER_CALLBACKS     0U /* TIM register callback disabled     */
+#define  USE_HAL_TIM_REGISTER_CALLBACKS     0U /* TIM register callback disabled    */
```
Comment (brunofalmeida) [inline Startup/STM32H7/stm32h7xx_hal_conf.h:216]: Should this be 1 to match the G4 configuration?

File: Src/drivers/STM32_tim/tim.c [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+TimFunc qs_timer;
+void (*qs_func)() = NULL;
+
+/* Function for TIM4 interrupt handler */
+void TIM4_IRQHandler(){
+	if(tim_get_IT_flag(&(qs_timer.handle), TIM_IT_UPDATE) != RESET){
+		tim_clear_IT_flag(&(qs_timer.handle), TIM_IT_UPDATE);
+		qs_func();
+	}
+}
```
Comment (brunofalmeida) [inline Src/drivers/STM32_tim/tim.c:431]: Could have something like this for all the timers so they can each have a saved callback function? Of course you would want to minimize code duplication as much as possible, so could define separate functions for `TIM4_IRQHandler()`, `TIM5_IRQHandler()`, etc. and have them call some common function with refactored logic. (If this doesn't make sense, let me know.)

File: Projects/STM32G474RETX/STM32G474RETX.ioc [hunk 8]
```diff
@@ -107,8 +107,8 @@ ProjectManager.MainLocation=Core/Src
 ProjectManager.NoMain=false
 ProjectManager.PreviousToolchain=
 ProjectManager.ProjectBuild=false
-ProjectManager.ProjectFileName=Firmware.ioc
-ProjectManager.ProjectName=Firmware
+ProjectManager.ProjectFileName=FIRMWARE.ioc
+ProjectManager.ProjectName=FIRMWARE
```
Comment (brunofalmeida) [inline Projects/STM32G474RETX/STM32G474RETX.ioc:111]: Not sure why this is in all caps. (was that the old CubeIDE project name?) Might want to just revert all changes to this file.

**Review by brunofalmeida — COMMENTED** [review]
General question: Did you find a way to make the timer callbacks more generic so that you can pass a function pointer to a callback when initializing one of the timers? The general use case is that some other code using a timer won't know which timer its callback will be connected to, so the function should not be forced to connect to a specific timer. If you're not quite sure how to accomplish this, ask Brytni or me.

Great work and your library is looking much improved, but I think it could still be improved further to be more modular as I mentioned.

File: Startup/STM32H7/stm32h7xx_hal_conf.h (hunk 6 above)
Comment (cam-rod) [reply Startup/STM32H7/stm32h7xx_hal_conf.h:216]: Setting the equivalent in the G4 file to `0U`. This was from when I was using the default callbacks, but since we're using our own, both should be disabled.

File: Src/drivers/STM32_tim/tim.h (hunk 5 above)
Comment (cam-rod) [reply Src/drivers/STM32_tim/tim.h:65]: These are optional commands for running calculations without doing it manually. They aren't needed for the timer driver to operate properly, so I updated the comment to make that clearer. I also added an equivalent `tim_calc_prescaler` for the H4.

File: Manual_Tests/STM32_tim/tim_test.c [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+volatile uint8_t led_toggle[] = {1,1,1,1,1,0,0,1,1,1,0,0,
+						  1,1,1,1,0,0,
+						  1,0,0,1,1,1,0,0,
+						  1,0,0,1,1,1,1,0,0,1,1,1,0,0,
+						  1,1,1,1,1,1,1,1,0,0,0,0,0,0}; // Each line is a letter and the trailing space or end of word.
+
+// Interrupt handler, advances at the end of each period
+void TIM2_IRQHandler(){
+	if(tim_get_IT_flag(&(timer.handle), TIM_IT_UPDATE)){
+		tim_clear_IT_flag(&(timer.handle), TIM_IT_UPDATE);
```
[the remark this answers is not in the document]
Comment (cam-rod) [reply Manual_Tests/STM32_tim/tim_test.c:30]: Added a function to the timer driver called `tim_check_flag` that handles this and returns 1 or 0 if the flag was raised. If you meant something else, let me know.

File: Src/drivers/STM32_tim/tim.c (hunk 4 above)
Comment (cam-rod) [reply Src/drivers/STM32_tim/tim.c:460]: I had included these since I wasn't sure if it was safe to assume the user had already called them, but since they're so general, I'll remove them.

File: Manual_Tests/STM32_tim/tim_test.c (hunk 2 above)
Comment (cam-rod) [reply Manual_Tests/STM32_tim/tim_test.c:32]: Asked Brytni, I'll update the file if she gets a change to add it, or I'll make a separate pull request for that file.

File: Manual_Tests/STM32_tim/tim_test.c [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+		if(led_toggle[time_unit]){
+			HAL_GPIO_TogglePin(GPIOA, GPIO_PIN_5);
+		}
+
+		time_unit++;
+		if(time_unit==54){ //Reset the cycle
+			time_unit=0;
```
Comment (cam-rod) [inline Manual_Tests/STM32_tim/tim_test.c:37]: [This](https://github.com/spacesys-finch/Firmware/pull/5#discussion_r506982377) makes sense, I should be able to implement this in a separate file to help keep the main timer driver clean. As for [this comment](https://github.com/spacesys-finch/Firmware/pull/5#pullrequestreview-511007930), did you mean creating a general function that will take a timer handle and function, and then setup the callback (for that specific timer) with the function provided?

If I understand correctly, the code highlighted would become a function pointer passed to this general function, which would then implement TIM4's interrupt handler.

File: Manual_Tests/STM32_tim/tim_test.c [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+#include <stn32h7xx_hal_gpio.h>
+#define LED_PORT GPIOB
+#define LED_PIN GPIO_PIN_7
+#endif
+
+static TimFunc timer;
+static GPIO_OUTPUT led_struct;
+volatile uint8_t time_unit = 0;
+volatile uint8_t led_toggle[] = {1,1,1,1,1,0,0,1,1,1,0,0,
+						  1,1,1,1,0,0,
```
Comment (BrytniR) [inline Manual_Tests/STM32_tim/tim_test.c:27]: Is there a reason there are uneven spaces here?

Comment (BrytniR) [reply Manual_Tests/STM32_tim/tim_test.c:27]: I mean the first line compared to the rest (unless this is just a github IDE problem)

Comment (cam-rod) [reply Manual_Tests/STM32_tim/tim_test.c:27]: image
Should've looked like this. Each line represents one of the letters in _FINCH_ that it flashes in Morse code.

**Comment by BrytniR** [comment]
If you try to build it, does it build with no errors?
(I just want to check if it's something on my side, because I'm getting errors like: Src/drivers/STM32_tim/tim.c:425:9: error: conflicting types for 'tim_check_flag'
 uint8_t tim_check_flag(TimFunc* timer, uint16_t flag){

File: Manual_Tests/STM32_tim/tim_test.c (hunk 2 above)
Comment (cam-rod) [reply Manual_Tests/STM32_tim/tim_test.c]: Added in #7, will be added to this file.

**Comment by cam-rod** [comment]
I hadn't updated the .h file, it'll be fixed in the next commit.

File: Manual_Tests/STM32_tim/tim_test.c (hunk 10 above)
Comment (cam-rod) [reply Manual_Tests/STM32_tim/tim_test.c]: Completed in [the latest commit](https://github.com/spacesys-finch/Firmware/commit/3abc1e4a2a750b105ec13d9601c8bba929431880).

**Comment by cam-rod** [comment]
I've completed all of the recommendations Bruno had suggested.

File: Manual_Tests/STM32_tim/tim_test.c [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+
+#include <STM32_tim/tim.h>
+
+#ifdef STM32G474xx
+#include <stm32g4xx_hal_gpio.h>
+#define LED_PORT GPIOA
+#define LED_PIN GPIO_PIN_5
+#elif defined(STM32H743xx)
+#include <stn32h7xx_hal_gpio.h>
```
Comment (brunofalmeida) [inline Manual_Tests/STM32_tim/tim_test.c:18]: Typo here ("stn")

File: Manual_Tests/STM32_tim/tim_test.c [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+ * tim_test.c
+ *
+ *  Created on: Jul 22, 2020
+ *      Author: Cameron Rodriguez
+ *
+ * Overview: A timer that uses LD2 to flash 'FINCH' in Morse code on the NUCLEO-G474RE. Based on https://visualgdb.com/tutorials/arm/stm32/timers/hal/
+ */
+
+
+#include <STM32_tim/tim.h>
```
Comment (brunofalmeida) [inline Manual_Tests/STM32_tim/tim_test.c:11]: I get an error here: `fatal error: STM32_tim/tim.h: No such file or directory`. Do you get the same error? Changing it to `#include <drivers/STM32_tim/tim.h>` fixes it for me.

File: Manual_Tests/STM32_tim/tim_test.c [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+#define LED_PORT GPIOA
+#define LED_PIN GPIO_PIN_5
+#elif defined(STM32H743xx)
+#include <stn32h7xx_hal_gpio.h>
+#define LED_PORT GPIOB
+#define LED_PIN GPIO_PIN_7
+#endif
+
+static TimFunc timer;
+static GPIO_OUTPUT led_struct;
```
Comment (brunofalmeida) [inline Manual_Tests/STM32_tim/tim_test.c:24]: I get an error here: `error: unknown type name 'GPIO_OUTPUT'`. Adding `#include <drivers/STM32_gpio/gpio.h>` fixes it for me.

File: Manual_Tests/STM32_tim/tim_test.c [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+volatile uint8_t time_unit = 0;
+volatile uint8_t led_toggle[] = {1,1,1,1,1,0,0,1,1,1,0,0,
+						  1,1,1,1,0,0,
+						  1,0,0,1,1,1,0,0,
+						  1,0,0,1,1,1,1,0,0,1,1,1,0,0,
+						  1,1,1,1,1,1,1,1,0,0,0,0,0,0}; // Each line is a letter and the trailing space or end of word.
+
+// Interrupt handler, advances at the end of each period
+void TIM2_IRQHandler(){
```
Comment (brunofalmeida) [inline Manual_Tests/STM32_tim/tim_test.c:33]: I get an error related to this function: `multiple definition of 'TIM2_IRQHandler'`. Since you defined this function (specifically the IRQ for TIM2) in the timer library itself, you would now just make this a generic callback function and call it anything you want (could be attached to any of the timers, which is the goal of what Brytni and I suggested you do).

File: Manual_Tests/STM32_tim/tim_test.c [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+ * Update events occur at each other time unit (300 ms or 20/3 Hz).
+ */
+int main(){
+	setup_tim();
+	led_struct = gpio_init_output(LED_PORT, LED_PIN);
+
+	tim_start(&timer);
+
+	return 0;
+}
```
Comment (brunofalmeida) [inline Manual_Tests/STM32_tim/tim_test.c:67]: Do you need to update this testing code to call `tim_setup_callback()` or does that automatically get called from your timer initialization functions?

File: Manual_Tests/STM32_tim/tim_test.c [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+volatile uint8_t time_unit = 0;
+volatile uint8_t led_toggle[] = {1,1,1,1,1,0,0,1,1,1,0,0,
+						  1,1,1,1,0,0,
+						  1,0,0,1,1,1,0,0,
+						  1,0,0,1,1,1,1,0,0,1,1,1,0,0,
+						  1,1,1,1,1,1,1,1,0,0,0,0,0,0}; // Each line is a letter and the trailing space or end of word.
+
+// Interrupt handler, advances at the end of each period
+void TIM2_IRQHandler(){
+	if(tim_check_flag(&timer, TIM_IT_UPDATE)){
```
Comment (brunofalmeida) [inline Manual_Tests/STM32_tim/tim_test.c:34]: Is this line now being taken care of in the library's IT handler functions?

File: Src/drivers/STM32_tim/tim_it_handlers.c [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+ *     Author: Cameron Rodriguez
+ *
+ * Overview: allows callbacks to be setup whenever an update interrupt occurs.
+ */
+
+#include "tim.h"
+void tim_run_callback(TIM_TypeDef *timer, uint8_t func_location);
+
+// Array of timers. Timers available: 1-7, 12-14 (H7 only), 15-17, 20 (G4 only)
+volatile void (*callback_array[14])() = {NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL};
```
Comment (brunofalmeida) [inline Src/drivers/STM32_tim/tim_it_handlers.c:13]: I suggest you call this `tim_callback_array` since it's a global variable and easy to conflict with other names from future libraries.

File: Src/drivers/STM32_tim/tim_it_handlers.c [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+/* This function runs the callback function of a given timer
+ *
+ * @param TIM_TypeDef *timer - the timer instance
+ * @param uint8_t func_location - location of the timer's callback function
+ */
+void tim_run_callback(TIM_TypeDef *timer, uint8_t func_location){
+	timer->SR = ~TIM_IT_UPDATE; // Clear flag
+	(*callback_array[func_location])();// Run the callback function
```
Comment (brunofalmeida) [inline Src/drivers/STM32_tim/tim_it_handlers.c:190]: Just in case something goes horribly wrong, I suggest you check that `func_location < 14` and `callback_array[func_location] != NULL` before running the function.

File: Src/drivers/STM32_tim/tim.c [hunk 20]
```diff
@@ -0,0 +1,478 @@
...
+TimFunc qs_timer;
+void (*qs_func)() = NULL;
+
+/* Function for TIM4 interrupt handler *
+void TIM4_IRQHandler(){
+	if(tim_get_IT_flag(&(qs_timer.handle), TIM_IT_UPDATE) != RESET){
+		tim_clear_IT_flag(&(qs_timer.handle), TIM_IT_UPDATE);
+		qs_func();
+	}
+} */
```
Comment (brunofalmeida) [inline Src/drivers/STM32_tim/tim.c:447]: Can get rid of this code since it's been replaced by the generic IT callbacks.

File: Src/drivers/STM32_tim/tim.c [hunk 21]
```diff
@@ -0,0 +1,478 @@
...
+	} else {
+		prescaler = (clk_frequency / target_frequency) - 1;
+		period = (clk_frequency / target_frequency / (prescaler+1)) - 1;
+		qs_timer = tim_config_base(TIM4, mode, prescaler, period);
+	}
+
+	tim_start(&qs_timer);
+
+	return qs_timer;
+}
```
Comment (brunofalmeida) [inline Src/drivers/STM32_tim/tim.c:478]: I'm not really familiar with what you and Brytni discussed, but is there a reason this quickstart timer specifically uses TIM4? Could it be made more generic to be able to use any of the timers?

Also, is this quickstart timer going to work with the new `tim_it_handlers.c` file? Does it need to call `tim_setup_callback()` to add the function to the callbacks array?

**Review by brunofalmeida — COMMENTED** [review]
I'm not sure if the compilation issues are just for me, but take a look and make sure you have the correct manual test file added to the CubeIDE project before compiling.

The restructured timer callbacks are looking really good.

File: Manual_Tests/STM32_tim/tim_test.c (hunk 13 above)
Comment (cam-rod) [reply Manual_Tests/STM32_tim/tim_test.c:11]: Not sure why that didn't appear on my system (double-checked the included paths and it should've raised an issue). Updated the include path.

File: Manual_Tests/STM32_tim/tim_test.c (hunk 17 above)
Comment (cam-rod) [reply Manual_Tests/STM32_tim/tim_test.c:34]: Yes, line 34 is the default. I included a message saying to comment out the lines if a different timer callback is needed.

File: Src/drivers/STM32_tim/tim_it_handlers.c (hunk 18 above)
Comment (cam-rod) [reply Src/drivers/STM32_tim/tim_it_handlers.c:13]: Ah, whoops, thanks for pointing it out.

File: Src/drivers/STM32_tim/tim.c (hunk 21 above)
Comment (cam-rod) [reply Src/drivers/STM32_tim/tim.c:478]: This was based around the older version of the callbacks when i was still figuring this out, so I locked it to one timer.

I forgot to update these components when I rebuilt the callbacks, so I'll switch this over as well as the test file.

**Comment by cam-rod** [comment]
Made all the suggested changes, and this version compiles correctly on my system. @brunofalmeida Is there some change you made to the build settings before you compiled the project? I'm trying to figure out why I didn't get any errors on my end originally.

**Comment by brunofalmeida** [comment]
I modified the project sources to use the `STM32_tim` manual test instead of the default `STM32_gpio` manual test in the project settings that are committed to Git. I was using the H7 project but it's the same thing for the G4 project if you are using that.

Screen Shot 2020-11-07 at 14 04 13

Note you are just supposed to change the project sources to include whichever manual test you are working on, but don't commit any of those CubeIDE project file changes to Git.

**Comment by brunofalmeida** [comment]
@cam-rod Are you sure you're using the new CubeIDE project structure with the project files under the `Projects` folder? You can use either the G4 or H7 project for development, just make sure before this branch is merged that is compiles properly for both the G4 and H7.

To be sure you're using the new project files, it's probably a good idea to close all projects in CubeIDE and reopen one of the new projects.

File: Src/drivers/STM32_tim/tim_it_handlers.c [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+/* This function runs the callback function of a given timer
+ *
+ * @param TIM_TypeDef *timer - the timer instance
+ * @param uint8_t func_location - location of the timer's callback function
+ */
+void tim_run_callback(TIM_TypeDef *timer, uint8_t func_location){
+	timer->SR = ~TIM_IT_UPDATE; // Clear flag
+	if(func_location>=0 && func_location<14){
```
Comment (brunofalmeida) [inline Src/drivers/STM32_tim/tim_it_handlers.c:190]: You should also be checking that the function pointer is not `NULL`. This is different from the value of `func_location` - I meant to check that the value in the array, i.e. `time_callback_array[func_location]` is not `NULL`.

**Comment by cam-rod** [comment]
@brunofalmeida Thanks for the pointers. Just tested the current commit for both projects, and they compiles without warnings or errors.

File: Src/drivers/STM32_tim/tim.c [hunk 23]
```diff
@@ -0,0 +1,477 @@
...
+} */
+
+/* Creates a Base timer on TIM4 that executes a void function at every update event.
+ *
+ * @param uint32_t clk_frequency - frequency at which the clock is operating
+ * @param uint32_t target_frequency - the frequency at which the function should execute (min. 1/65536 of clk_frequency)
+ * @param void (*f)() - the void function to execute
+ * @return TimFunc - the loaded timer struct
+ */
+TimFunc tim_quickstart_void_function(uint32_t clk_frequency, uint32_t target_frequency, void (*qs_function)()){
```
Comment (brunofalmeida) [inline Src/drivers/STM32_tim/tim.c:455]: Is this going to be updated to be generic rather than being tied to `TIM4` or is there a reason to stick with this current design?

**Review by brunofalmeida — COMMENTED** [review]
It all compiles fine for me now with the H7 project without any changes.

**Review by brunofalmeida — COMMENTED** [review]
I don't have any more comments, seems all good to me. Obviously you need to test it on the board to see if it works, and Brytni may have more comments.

**Comment by cam-rod** [comment]
Replaced with #15.
